libraries/appbase/application.cpp

```diff
@@ -18,104 +18,140 @@ using bpo::options_description;
+         bfs::path configs_dir = actual_data_dir / "configs/";
+         if (bfs::exists(configs_dir) == false) {
+            bfs::create_directory(configs_dir);
+         }
+
+         bfs::rename(deprecated_default_config_path, configs_dir / cfg_config.as<bfs::path>());
       }
-      const variables_map*    _options = nullptr;
-      options_description     _app_options;
-      options_description     _cfg_options;
-      variables_map           _args;
+   }
+}*/
 
-      bfs::path               _data_dir;
-};
+/**
+          * @brief Writes options_desc data into cfg_file file
+          *
+          * @param options_desc
+          * @param cfg_file
+          * @return created config file absolute path
+          */
+bfs::path write_default_config( const options_description& options_desc, const bfs::path& cfg_file, std::string suffix = "" ) {
+   bfs::path _cfg_file = cfg_file;
+   if( suffix != "" ){
+      _cfg_file += std::string(".") + suffix;
+   }
 
-application::application()
-:my(new application_impl()){
-   io_serv = std::make_shared<boost::asio::io_service>();
-}
+   bfs::path config_file_name = app_factory().data_dir / "configs" / _cfg_file ;
 
-application::~application() { }
+   if( _cfg_file.is_absolute() == true ) {
+      config_file_name = _cfg_file;
+   }
 
-void application::startup() {
-   for (const auto& plugin : initialized_plugins)
-      plugin->startup();
-}
+   if(bfs::exists(config_file_name) == true) {
+      return config_file_name;
+   }
+
+   if(bfs::exists(config_file_name.parent_path()) == false) {
+      bfs::create_directories(config_file_name.parent_path());
+   }
 
-application& application::instance( bool reset ) {
-   static application* _app = new application();
-   if( reset )
+   std::ofstream out_cfg( bfs::path(config_file_name).make_preferred().string());
+   for(const boost::shared_ptr<bpo::option_description> od : options_desc.options())
    {
-      delete _app;
-      _app = new application();
+      if(!od->description().empty())
+         out_cfg << "# " << od->description() << "\n";
+      boost::any store;
+      if(!od->semantic()->apply_default(store))
+         out_cfg << "# " << od->long_name() << " = \n";
+      else {
+         auto example = od->format_parameter();
+         if(example.empty())
+            // This is a boolean switch
+            out_cfg << od->long_name() << " = " << "false\n";
+         else {
+            // The string is formatted "arg (=<interesting part>)"
+            example.erase(0, 6);
+            example.erase(example.length()-1);
+            out_cfg << od->long_name() << " = " << example << "\n";
+         }
+      }
+      out_cfg << "\n";
    }
-   return *_app;
+   out_cfg.close();
+
+   return config_file_name;
 }
-application& app() { return application::instance(); }
-application& reset() { return application::instance( true ); }
 
+const static string _public_net_chain_id = "1a058d1a89aff240ab203abe8a429d1a1699c339032a87e70e01022842a98324";
```

Comment:
Will this work with testnet?

Reply:
Create separate config file for testnet (e.g. config.ini.testnet) and put reference to the net in main config.ini.
This exists only to make sure that running the binary without explicitly configuring any network results in starting the mainnet.